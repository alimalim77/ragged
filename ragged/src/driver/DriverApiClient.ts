export class DriverApiClient {
    post(url: string, request: RequestInit) {
        request.method = "POST";
        return fetch(url, request);
    }
}