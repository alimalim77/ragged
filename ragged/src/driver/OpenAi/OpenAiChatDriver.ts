import { Logger } from "../../logger/Logger";
import { DriverApiClient } from "../DriverApiClient";
import { JsonParseError, FetchRequestFailedError } from "../DriverErrors";

// ============ types ============

type OpenAiChatDriverConfig = {
    apiKey: string | undefined;
    rootUrl: string;
}

type ChatCompletionParams = {
    model: string;
    messages: {
        role: string;
        content: string;
    }[]
}


// ============ utils ============

const buildDefaultConfig = (): OpenAiChatDriverConfig => {
    return {
        apiKey: undefined,
        rootUrl: "https://api.openai.com/v1/chat/completions"
    }
}

/**
 * @private
 * This is a driver for the OpenAI Chat API. Most devs will never instantiate this class directly, and don't
 * need to know about it.
 */
export class OpenAiChatDriver {
    private logger: Logger = new Logger('OpenAiChatDriver');
    private config: OpenAiChatDriverConfig;

    constructor(private driverApiClient: DriverApiClient, config: Partial<OpenAiChatDriverConfig> = {}) {
        this.config = { ...buildDefaultConfig(), ...config };
    }

    async chatCompletion(params: ChatCompletionParams) {
        let body: string;

        try {
            body = JSON.stringify(params);
        } catch (e) {
            const message = "Failed to stringify request JSON for OpenAI API.";
            this.logger.error(message, e);
            throw new JsonParseError(message, e);
        }

        const response = await this.driverApiClient.post(
            this.config.rootUrl,
            {
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${this.config.apiKey}`
                },
                body
            }
        )

        if (response.status < 200 || response.status >= 300) {
            const err = new FetchRequestFailedError(`Request to OpenAI failed with status code ${response.status}.`, response)
            this.logger.error(err);
            throw err;
        }

        let json: any;
        try {
            json = response.json();
        } catch (e) {
            const err = new JsonParseError("Failed to parse response JSON from OpenAI API.", e);
            this.logger.error(err);
            throw err;
        }

        return json;
    }
}