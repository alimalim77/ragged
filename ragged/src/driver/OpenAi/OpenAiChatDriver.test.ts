import { DriverApiClient } from "../DriverApiClient";
import { OpenAiChatDriver } from "./OpenAiChatDriver";
import nock from 'nock';

describe("OpenAiChatDriver", () => {
    beforeEach(() => {
        nock('https://api.openai.com')
            .post('/v1/chat/completions')
            .reply(200, {
                "choices": [
                    {
                        "finish_reason": "stop",
                        "index": 0,
                        "logprobs": null,
                        "text": "I am a chatbot."
                    }
                ]
            });
    })
    it("should perform a chat completion", async () => {
        // Arrange
        const config = {
            apiKey: process.env.OPENAI_API_KEY
        }

        const driverApiClient = new DriverApiClient();
        const driver = new OpenAiChatDriver(driverApiClient, config);

        // Act
        const result = await driver.chatCompletion({
            model: "gpt-3.5-turbo",
            messages: [
                { role: "system", content: "You are a chatbot." },
                { role: "user", content: "What are you?" }
            ]
        });

        console.log(JSON.stringify(result));
    })
})