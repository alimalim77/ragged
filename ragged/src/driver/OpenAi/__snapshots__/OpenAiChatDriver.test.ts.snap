// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenAiChatDriver should perform a chat completion 1`] = `
{
  "choices": [
    {
      "finish_reason": "stop",
      "index": 0,
      "logprobs": null,
      "message": {
        "content": "I am a chatbot designed to assist you with information and answer your questions to the best of my abilities. How can I help you today?",
        "role": "assistant",
      },
    },
  ],
  "created": 1717283755,
  "id": "chatcmpl-9VSF1LRQZj65RIITve1CqGB9VGALo",
  "model": "gpt-3.5-turbo-0125",
  "object": "chat.completion",
  "system_fingerprint": null,
  "usage": {
    "completion_tokens": 29,
    "prompt_tokens": 21,
    "total_tokens": 50,
  },
}
`;
